interfaces/prompting/requestrules/requestrules.go

```diff
@@ -452,31 +463,57 @@ func (rdb *RuleDB) addRulePermissionToTree(rule *Rule, permission string) []prom
 	var conflicts []prompting_errors.RuleConflict
 
 	addVariant := func(index int, variant patterns.PatternVariant) {
+		variantStr := variant.String()
+		existingEntry, exists := permVariants.VariantEntries[variantStr]
+		if !exists {
+			newVariantEntries[variantStr] = variantEntry{
+				Variant: variant,
+				Outcome: rule.Outcome,
+				RuleIDs: map[prompting.IDType]bool{rule.ID: true},
+			}
+			return
+		}
+		if existingEntry.Outcome != rule.Outcome {
+			// Outcomes mismatch, so if there are any unexpired rules, these
+			// all conflict with the new rule. If all rules are expired, this
+			// is fine, and we can replace the existing entry with a new one.
+			for id := range existingEntry.RuleIDs {
+				if rdb.isRuleWithIDExpired(id, rule.Timestamp) {
+					expiredRules[id] = true
+					continue
+				}
+				// Conflicting non-expired rule
+				conflicts = append(conflicts, prompting_errors.RuleConflict{
+					Permission:    permission,
+					Variant:       variantStr,
+					ConflictingID: id.String(),
+				})
+			}
+		}
+		if len(conflicts) > 0 {
+			// If there are any conflicts (for this variant or others), all
+			// changes will be discarded, so don't bother building the new
+			// variant entry
+			return
+		}
 		newEntry := variantEntry{
```

Comment:
setting this partially up could be done before the previous "if", same with filling newVariantEntries[variantStr]

Reply:
Ah I see, yes the existing logic is convoluted, relies on lots of checks and early returns. I've cleaned this up in the latest commit. Moved this bit up before the for loop, and consolidated the two for loops together with fewer unnecessary checks.